packages/backend/src/nest/libp2p/process-in-chunks.service.ts

```diff
@@ -27,59 +28,92 @@ export class ProcessInChunksService<T> extends EventEmitter {
     this.logger(`Initializing process-in-chunks.service with peers ${JSON.stringify(data, null, 2)}`)
     this.processItem = processItem
     this.chunkSize = chunkSize
-    this.taskQueue = fastq(this, this.processOneItem, this.chunkSize)
-    this.updateData(data)
-    this.addToTaskQueue()
+    this.taskQueue = fastq.promise(this, this.processOneItem, this.chunkSize)
+    this.isActive = true
+    this.updateQueue(data)
   }
 
-  public updateData(items: T[]) {
-    this.logger(`Updating data with ${items.length} items`)
-    this.taskQueue.pause()
-    items.forEach(item => this.data.add(item))
-    this.addToTaskQueue()
+  public updateQueue(items: T[]) {
+    this.logger(`Adding ${items.length} items to the task queue`)
+    items.forEach(item => this.addToTaskQueue(item))
   }
 
-  private addToTaskQueue() {
-    this.logger(`Adding ${this.data.size} items to the task queue`)
-    for (const item of this.data) {
-      if (item) {
-        this.logger(`Adding data ${item} to the task queue`)
-        this.data.delete(item)
-        try {
-          this.taskQueue.push({ data: item, tries: 0 } as ProcessTask<T>)
-        } catch (e) {
-          this.logger.error(`Error occurred while adding new task for item ${item} to the queue`, e)
-          this.data.add(item)
-        }
+  private async addToTaskQueue(task: ProcessTask<T>): Promise<void>
+  private async addToTaskQueue(item: T): Promise<void>
+  private async addToTaskQueue(itemOrTask: T | ProcessTask<T>): Promise<void> {
+    if (!itemOrTask) {
+      this.logger.error('Item/task is null or undefined, skipping!')
+      return
+    }
+
+    let task: ProcessTask<T>
+    if ((itemOrTask as ProcessTask<T>).taskId != null) {
+      task = itemOrTask as ProcessTask<T>
+    } else {
+      this.logger(`Creating new task for ${itemOrTask}`)
+      task = { data: itemOrTask as T, tries: 0, taskId: randomUUID() }
+    }
+
+    if (!this.isActive) {
+      this.logger(
+        'ProcessInChunksService is not active, adding tasks to the dead letter queue!\n\nWARNING: You must call "resume" on the ProcessInChunksService to process the dead letter queue!!!'
+      )
+      this.deadLetterQueue.push(task)
+      return
+    }
+
+    this.logger(`Adding task ${task.taskId} with data ${task.data} to the task queue`)
+    try {
+      const success = await this.taskQueue.push(task)
+      if (!success) {
+        this.logger(`Will try to re-attempt task ${task.taskId} with data ${task.data}`)
+        await this.taskQueue.push({ ...task, tries: task.tries + 1 })
```

Comment:
Ideally we would retry differently depending on the error but figuring out all of the errors that are possible with libp2p that amount to "we couldn't connect because we haven't been able to yet" or "this peer is offline and we won't connect right now no matter what" to filter those out and only retry multiple times when its something that would keep us from adding the peer to the peerstore and auto-dialing.